### PR TITLE
Use std::to_string for Convert::ToString()

### PR DESCRIPTION
**Pull request by mcktr**
This uses `std::to_string` in `Convert::ToString` instead of
`boost::lexical_cast<std::string>`. This also adds a special handling for
a char array.

refs #5842

### CODE REVIEW

File: lib/base/configtype.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ void ConfigType::RegisterObject(const ConfigObject::Ptr& object)
 			auto *type = dynamic_cast<Type *>(this);
 
 			BOOST_THROW_EXCEPTION(ScriptError("An object with type '" + type->GetName() + "' and name '" + name + "' already exists (" +
-				Convert::ToString(it->second->GetDebugInfo()) + "), new declaration: " + Convert::ToString(object->GetDebugInfo()),
+				it->second->GetDebugInfo().Path + "), new declaration: " + object->GetDebugInfo().Path,
```
Comment (dnsmichi) [inline lib/base/configtype.cpp]: Why the change here?

**Comment by mcktr** [comment]
I just learned that the change in `configtype.cpp` does not work fully, since the line and column information are missing in the log exception message.

Without this PR:

```
[2019-02-05 18:31:55 +0100] critical/config: Error: An object with type 'Service' and name 'metis!ping4' already exists (in /usr/local/icinga2/etc/icinga2/conf.d/services.conf: 34:1-34:21), new declaration: in /usr/local/icinga2/etc/icinga2/conf.d/services.conf: 26:1-26:21
``` 

With this PR:

```
[2019-02-05 18:08:47 +0100] critical/config: Error: An object with type 'Service' and name 'metis!ping4' already exists (/usr/local/icinga2/etc/icinga2/conf.d/services.conf), new declaration: /usr/local/icinga2/etc/icinga2/conf.d/services.conf
```

I now came up with a new idea to create a member function `ToString()` in the struct `DebugInfo` to convert it.

The change originally was introduced because otherwise there is a build error.

```
[  5%] Building CXX object lib/base/CMakeFiles/base.dir/convert.cpp.o
In file included from /home/michael/dev/icinga/icinga2/lib/base/configtype.cpp:21:
/home/michael/dev/icinga/icinga2/lib/base/convert.hpp:86:10: error: no matching function for call to 'to_string'
                return std::to_string(val);
                       ^~~~~~~~~~~~~~
/home/michael/dev/icinga/icinga2/lib/base/configtype.cpp:57:14: note: in instantiation of function template specialization 'icinga::Convert::ToString<icinga::DebugInfo>' requested here
                                Convert::ToString(it->second->GetDebugInfo()) + "), new declaration: " + Convert::ToString(object->GetDebugInfo()),
                                         ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6405:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'int' for 1st argument
  to_string(int __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6410:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'unsigned int' for 1st argument
  to_string(unsigned __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6416:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'long' for 1st argument
  to_string(long __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6421:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'unsigned long' for 1st argument
  to_string(unsigned long __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6427:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'long long' for 1st argument
  to_string(long long __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6433:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'unsigned long long' for 1st argument
  to_string(unsigned long long __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6439:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'float' for 1st argument
  to_string(float __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6448:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'double' for 1st argument
  to_string(double __val)
  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:6457:3: note: candidate function not viable: no known conversion from 'const icinga::DebugInfo' to 'long double' for 1st argument
  to_string(long double __val)
  ^
1 error generated.
```

**Comment by dnsmichi** [comment]
I'm wondering why the configtype string worked previously. Can you do a test with boost::lexical_cast and DebugInfo only? We may miss some specific type casts here.

**Comment by mcktr** [comment]
Did not work: 

```
In file included from /usr/include/boost/lexical_cast.hpp:32:
/usr/include/boost/lexical_cast/try_lexical_convert.hpp:95:22: error: invalid application of 'sizeof' to an incomplete type 'icinga::DebugInfo'
                     sizeof(Source) == sizeof(Target) &&
                     ^~~~~~~~~~~~~~
/usr/include/boost/lexical_cast/try_lexical_convert.hpp:166:32: note: in instantiation of template class 'boost::detail::is_xchar_to_xchar<std::__cxx11::basic_string<char>, icinga::DebugInfo>' requested here
                boost::detail::is_xchar_to_xchar<Target, src >::value ||
                               ^
/usr/include/boost/lexical_cast.hpp:41:41: note: in instantiation of function template specialization 'boost::conversion::detail::try_lexical_convert<std::__cxx11::basic_string<char>, icinga::DebugInfo>' requested here
        if (!boost::conversion::detail::try_lexical_convert(arg, result)) {
                                        ^
/home/michael/dev/icinga/icinga2/lib/base/convert.hpp:91:17: note: in instantiation of function template specialization 'boost::lexical_cast<std::__cxx11::basic_string<char>, icinga::DebugInfo>' requested here
                return boost::lexical_cast<std::string>(val);
                              ^
/home/michael/dev/icinga/icinga2/lib/base/object.hpp:40:8: note: forward declaration of 'icinga::DebugInfo'
struct DebugInfo;
       ^
In file included from /home/michael/dev/icinga/icinga2/lib/base/convert.cpp:20:
In file included from /home/michael/dev/icinga/icinga2/lib/base/convert.hpp:25:
In file included from /usr/include/boost/lexical_cast.hpp:31:
/usr/include/boost/lexical_cast/bad_lexical_cast.hpp:92:53: error: 'typeid' of incomplete type 'icinga::DebugInfo'
            boost::throw_exception(bad_lexical_cast(typeid(S), typeid(T)));
                                                    ^
/usr/include/boost/lexical_cast.hpp:42:40: note: in instantiation of function template specialization 'boost::conversion::detail::throw_bad_cast<icinga::DebugInfo, std::__cxx11::basic_string<char> >' requested here
            boost::conversion::detail::throw_bad_cast<Source, Target>();
                                       ^
/home/michael/dev/icinga/icinga2/lib/base/convert.hpp:91:17: note: in instantiation of function template specialization 'boost::lexical_cast<std::__cxx11::basic_string<char>, icinga::DebugInfo>' requested here
                return boost::lexical_cast<std::string>(val);
                              ^
/home/michael/dev/icinga/icinga2/lib/base/object.hpp:40:8: note: forward declaration of 'icinga::DebugInfo'
struct DebugInfo;
       ^
2 errors generated.
```

But this lead me in the right direction. In `debuginfo.cpp` we overload the operator `<<` which we can use for the cast from `DebugInfo` to `std::string`. I pushed a new commit to implement this accordingly.

**Comment by dnsmichi** [comment]
Ok, thanks, I'll have a look into this.

**Comment by dnsmichi** [comment]
I've been reading "Boost C++ Application Development Cookbook - Second Edition: Recipes to simplify your application development" over the weekend.

Since our focus shifts a little to Boost again, I've also evaluated how boost::lexical_cast works in contrast with custom C++11 code. In this specific regard, it already optimizes things in the way we need and as such we shouldn't change the current behavior.

Sorry about that.

Cheers,
Michael
